pkg/ebpf/tcp_dns_layer.go

```diff
@@ -0,0 +1,52 @@
+// +build linux_bpf
+
+package ebpf
+
+import (
+	"encoding/binary"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+)
+
+// source: https://github.com/weaveworks/scope/blob/master/probe/endpoint/dns_snooper.go
+// Gopacket doesn't provide direct support for DNS over TCP, see https://github.com/google/gopacket/issues/236
+type tcpWithDNSSupport struct {
+	tcp layers.TCP
+}
+
+func (m *tcpWithDNSSupport) DecodeFromBytes(data []byte, df gopacket.DecodeFeedback) error {
+	return m.tcp.DecodeFromBytes(data, df)
+}
+
+func (m *tcpWithDNSSupport) CanDecode() gopacket.LayerClass { return m.tcp.CanDecode() }
+
+// Determine if a TCP segment contains a full DNS message (i.e. not fragmented)
+func (m *tcpWithDNSSupport) hasSelfContainedDNSPayload() bool {
+	payload := m.tcp.LayerPayload()
+	if len(payload) < 2 {
+		return false
+	}
+
+	// Assume it's a self-contained DNS message if the Length field
+	// matches the length of the TCP segment
+	dnsLengthField := binary.BigEndian.Uint16(payload)
+	return int(dnsLengthField) == len(payload)-2
+}
+
+func (m *tcpWithDNSSupport) NextLayerType() gopacket.LayerType {
+	// TODO: deal with TCP fragmentation and out-of-order segments
+	if (m.tcp.SrcPort == 53 || m.tcp.DstPort == 53) && m.hasSelfContainedDNSPayload() {
+		return layers.LayerTypeDNS
+	}
+	return m.tcp.NextLayerType()
+}
+
+func (m *tcpWithDNSSupport) LayerPayload() []byte {
+	payload := m.tcp.LayerPayload()
+	if len(payload) > 1 && (m.tcp.SrcPort == 53 || m.tcp.DstPort == 53) {
```

Comment:
Do we need to rely on the port here?  I get that port 53 is standard DNS but I'm wondering if there are weird internal corp setups that would use a different port.

I don't know enough about it to say one way or the other

Reply:
This is aligned with the check that we currently have in the lower layer: https://github.com/DataDog/datadog-agent/blob/master/pkg/ebpf/c/tracer-ebpf.c#L882. 

Now that I think about it, this check is redundant since only packets with source port `53` will reach this `Snooper` anyway. I will remove it. 

Whether we want to continue to restrict ourselves to port `53` in the lower layer is something I don't know much about either. I am guessing this was something that was discussed during the implementation. @p-lambert , do you have any thoughts on this?

My understanding is that without relying on `src_port` being `53`, there is no easy way to limit ourselves to DNS responses.

Reply:
Yeah, if we start relying on things other than the port number we'll need a more sophisticated socket filter (based on the packet payload), so it's not something worth to be "proactive" about, unless we validate a lot customers have exotic setups.